stixcore/products/level0/science.py

```diff
@@ -266,14 +266,253 @@ def is_datasource_for(cls, *, service_type, service_subtype, ssid, **kwargs):
+                in zip(e_min, e_max, energy_unit)]
+
+        control['energy_bin_mask'] = np.full((1, 32), False, np.ubyte)
+        control['energy_bin_mask'][:, cids] = True
+
+        dl_energies = np.array([[ENERGY_CHANNELS[ch].e_lower for ch in chs]
+                                + [ENERGY_CHANNELS[chs[-1]].e_upper] for chs in cids][0])
+
+        sci_energies = np.hstack([[ENERGY_CHANNELS[ch].e_lower for ch in range(32)],
+                                  ENERGY_CHANNELS[31].e_upper])
+        ind = 0
+        for nt in num_times:
+            e_ch_start = 0
+            e_ch_end = counts.shape[1]
+            if dl_energies[0] == 0:
+                full_counts[ind:ind + nt, 0] = counts[ind:ind + nt, 0]
+                full_counts_var[ind:ind + nt, 0] = counts_var[ind:ind + nt, 0]
+                e_ch_start = 1
+            if dl_energies[-1] == np.inf:
+                full_counts[ind:ind + nt, -1] = counts[ind:ind + nt, -1]
+                full_counts_var[ind:ind + nt, -1] = counts[ind:ind + nt, -1]
+                e_ch_end -= 1
+
+            torebin = np.where((dl_energies >= 4.0) & (dl_energies <= 150.0))
+            full_counts[ind:ind + nt, 1:-1] = np.apply_along_axis(
+                rebin_proportional, 1, counts[ind:ind + nt, e_ch_start:e_ch_end],
+                dl_energies[torebin], sci_energies[1:-1])
+
+            full_counts_var[ind:ind + nt, 1:-1] = np.apply_along_axis(
+                rebin_proportional, 1, counts_var[ind:ind + nt, e_ch_start:e_ch_end],
+                dl_energies[torebin], sci_energies[1:-1])
+
+            ind += nt
+
+        if counts.sum() != full_counts.sum():
+            raise ValueError('Original and reformatted count totals do not match')
+
+        delta_time = (np.array(packets.get_value('NIX00441'), np.uint16)) * 0.1 * u.s
+        closing_time_offset = (np.array(packets.get_value('NIX00269'), np.uint16)) * 0.1 * u.s
+
+        # TODO incorporate into main loop above
+        centers = []
+        deltas = []
+        last = 0
+        for i, nt in enumerate(num_times):
+            edge = np.hstack(
+                [delta_time[last:last + nt], delta_time[last + nt - 1] + closing_time_offset[i]])
+            delta = np.diff(edge)
+            center = edge[:-1] + delta / 2
+            centers.append(center)
+            deltas.append(delta)
+            last = last + nt
+
+        centers = np.hstack(centers)
+        deltas = np.hstack(deltas)
+
+        # Data
+        data = Data()
+        data['time'] = DateTime(control["time_stamp"][0], 0).as_float() + centers
+        data['timedel'] = deltas
+
+        triggers = packets.get_value('NIX00267')
+        triggers_var = packets.get_value('NIX00267', attr='error')
+
+        data['triggers'] = triggers
+        data['triggers_err'] = np.sqrt(triggers_var)
+        data['counts'] = full_counts * u.ct
+        data['counts_err'] = np.sqrt(full_counts_var) * u.ct
+        data['control_index'] = 0
+
```

Comment:
we should store the skm values used for the decompression:

`control['compression_scheme_triggers_skm'], \     control['compression_scheme_vis_skm'].meta = \    _get_compression_scheme(packets, 'NIX00267')`  also for counts